src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceRunner.java

```diff
@@ -0,0 +1,286 @@
+
+    private static DatasourceRunner INSTANCE;
+
+    /**
+     * Return a singleton job runner instance
+     * @return job runner
+     */
+    public static DatasourceRunner getJobRunnerInstance() {
+        if (INSTANCE != null) {
+            return INSTANCE;
+        }
+        synchronized (DatasourceRunner.class) {
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
+            INSTANCE = new DatasourceRunner();
+            return INSTANCE;
+        }
+    }
+
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private Client client;
+    private TimeValue timeout;
+    private int indexingBulkSize;
+
+    private DatasourceRunner() {
+        // Singleton class, use getJobRunner method instead of constructor
+    }
+
+    /**
+     * Set cluster service
+     * @param clusterService the cluster service
+     */
+    public void setClusterService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+
+        timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(clusterService.getSettings());
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, it -> timeout = it);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(clusterService.getSettings());
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, it -> indexingBulkSize = it);
+    }
+
+    /**
+     * Set thread pool
+     * @param threadPool the thread pool
+     */
+    public void setThreadPool(ThreadPool threadPool) {
+        this.threadPool = threadPool;
+    }
+
+    /**
+     * Set client
+     * @param client the client
+     */
+    public void setClient(Client client) {
+        this.client = client;
+    }
+
+    /**
+     * Update GeoIP data
+     *
+     * Lock is used so that only one of nodes run this task.
+     * Lock duration is 1 hour to avoid refreshing. This is okay because update interval is 1 day minimum.
+     *
+     * @param jobParameter job parameter
+     * @param context context
+     */
+    @Override
+    public void runJob(ScheduledJobParameter jobParameter, JobExecutionContext context) {
+        log.info("Update job started for a datasource[{}]", jobParameter.getName());
+        if (!(jobParameter instanceof Datasource)) {
+            throw new IllegalStateException(
+                "Job parameter is not instance of DatasourceUpdateJobParameter, type: " + jobParameter.getClass().getCanonicalName()
+            );
+        }
+
+        if (clusterService == null) {
+            throw new IllegalStateException("ClusterService is not initialized.");
+        }
+
+        if (threadPool == null) {
+            throw new IllegalStateException("ThreadPool is not initialized.");
+        }
+
+        if (client == null) {
+            throw new IllegalStateException("Client is not initialized.");
+        }
```

Comment:
These all can be avoided if you use a AllArgsConstructor and check everything parameters to be null in that function. Doing these null checks here pollutes the function.

Reply:
AllArgsConstructor cannot be used here because of the way how job scheduler create an instance.
Let me think what I can do here.